core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -384,15 +385,36 @@ private[spark] class MemoryStore(
     }
   }
 
+  private def maybeReleaseResources(entry: MemoryEntry[_]): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
+      case DeserializedMemoryEntry(objs: Array[Any], _, _) => maybeCloseValues(objs)
```

Comment:
As I know, broadcasted variables can be serialized on disk too (`BlockManager.doPutIterator`). In the case, seems `AutoCloseable` broadcasted variables won't hit this release logic.

Reply:
I wouldn't expect a never-deserialized Memory Entry to be closed, as it was never really instantiated to begin with - so if it _only_ lands on disk, I think that's reasonable (as the variable in question would never have had a chance to allocate anything either).

Reply:
Ah- ok, I see where the issue is. So in this case you may have a deserialized instance but the memory store is full, so it fails to be put. Now we've got a live, deserialized object not in MemoryStore. Thanks for catching this. It looks like this case could be handled in `MemoryStore.putIteratorAsValues` when the `putIterator` call fails, which would handle several cases in `BlockManager` where we try (and fail) to put deserialized values, but it means a check for potential `AutoClosable` values any time we fail to put into `MemoryStore`, and I'm not sure of the performance impact of this.

Reply:
Actually, digging further, there's other places where we may deserialize an object from the disk store and never put it into the memory store - it seems like punting on a guarantee that your AutoClosable object is closed and making this a best-effort thing when calling `BlockManager.removeBroadcast` (which is how we used it in the case that caused us to put together this PR) may make the most sense. It'll still be better than depending on GC and a finalizer to get the resources cleaned up when the driver can call `Broadcast#destroy` but we can document it as a best practice to have one just in case the close() call doesn't happen due to edge cases.